src/Compilers/Core/Portable/PortableShim.cs

```diff
@@ -0,0 +1,453 @@
+
+            internal static readonly object ReadWriteBitwiseOrDelete = Enum.ToObject(Type, 3 | 4);
+        }
+
+        internal static class FileOptions
+        {
+            internal const string TypeName = "System.IO.FileOptions";
+
+            internal static readonly Type Type = GetTypeFromEither(
+                contractName: $"{TypeName}, {System_IO_FileSystem_Name}",
+                desktopName: TypeName);
+
+            internal static readonly object None = Enum.ToObject(Type, 0);
+
+            internal static readonly object Asynchronous = Enum.ToObject(Type,  1073741824);
+        }
+
+        internal static class SearchOption
+        {
+            internal const string TypeName = "System.IO.SearchOption";
+
+            internal static readonly Type Type = GetTypeFromEither(
+                contractName: $"{TypeName}, {System_IO_FileSystem_Name}",
+                desktopName: TypeName);
+
+            internal static readonly object TopDirectoryOnly = Enum.ToObject(Type, 0);
+
+            internal static readonly object AllDirectories = Enum.ToObject(Type, 1);
+        }
+
+        internal static class FileStream
+        {
+            internal const string TypeName = "System.IO.FileStream";
+
+            internal static readonly Type Type = GetTypeFromEither(
+                contractName: $"${TypeName}, ${System_IO_FileSystem_Name}",
+                desktopName: TypeName);
+
+            internal static readonly PropertyInfo Name = Type
+                .GetTypeInfo()
+                .GetDeclaredProperty(nameof(Name));
+
+            private static ConstructorInfo s_Ctor_String_FileMode_FileAccess_FileShare = Type
+                .GetTypeInfo()
+                .GetDeclaredConstructor(paramTypes: new[] { typeof(string), FileMode.Type, FileAccess.Type, FileShare.Type });
+
+            private static ConstructorInfo s_Ctor_String_FileMode_FileAccess = Type
+                .GetTypeInfo()
+                .GetDeclaredConstructor(paramTypes: new[] { typeof(string), FileMode.Type, FileAccess.Type });
+
+            private static ConstructorInfo s_Ctor_String_FileMode = Type
+                .GetTypeInfo()
+                .GetDeclaredConstructor(paramTypes: new[] { typeof(string), FileMode.Type });
+
+            private static ConstructorInfo s_Ctor_String_FileMode_FileAccess_FileShare_Int32_FileOptions = Type
+                .GetTypeInfo()
+                .GetDeclaredConstructor(paramTypes: new[] { typeof(string), FileMode.Type, FileAccess.Type, FileShare.Type, typeof(int), FileOptions.Type });
+
+            public static Stream Create(string path, object mode)
+            {
+                try
+                {
+                    return (Stream)s_Ctor_String_FileMode.Invoke(new[] { path, mode });
+                }
+                catch (TargetInvocationException e)
+                {
+                    throw e.InnerException;
```

Comment:
Use ExceptionDispatchInfo instead of throwing these directly so that the previous stack isn't overridden.